source/Mlos.Python/mlos/Optimizers/OptimizationProblem.py

```diff
@@ -99,20 +105,47 @@ def __init__(
 
         # Fit functions / surrogate models will be fed features consisting of both context and parameters.
         # Thus, the feature space is comprised of both context and parameters.
+        has_context = self.context_space is not None
         self.feature_space = SimpleHypergrid(
             name="features",
             dimensions=[
-                CategoricalDimension(name="contains_parameters", values=[True]),
-                CategoricalDimension(name="contains_context", values=[self.context_space is not None])
+                CategoricalDimension(name="contains_context", values=[has_context])
             ]
         ).join(
             subgrid=self.parameter_space,
-            on_external_dimension=CategoricalDimension(name="contains_parameters", values=[True])
-        ).join(
-            subgrid=self.context_space,
-            on_external_dimension=CategoricalDimension(name="contains_context", values=[True])
+            on_external_dimension=CategoricalDimension(name="contains_context", values=[has_context])
+        )
+        if has_context:
+            self.feature_space = self.feature_space.join(
+                subgrid=self.context_space,
+                on_external_dimension=CategoricalDimension(name="contains_context", values=[True])
         )
 
+    def construct_feature_dataframe(self, config_values, context_values=None, product=False):

```

Comment:
Putting this into the Optimization problem seems the right place. wdyt?

Reply:
It violates the Law of Demeter a little bit, but we can deal with it in some subsequent PR.

Reply:
where would you put it?